pkg/controller/podautoscaler/horizontal.go

```diff
@@ -652,15 +689,96 @@ func (a *HorizontalController) reconcileAutoscaler(hpav1Shared *autoscalingv1.Ho
+	ScaleDownConstraint autoscalingv2.HPAScaleConstraintValue
+	MinReplicas         *int32
+	MaxReplicas         int32
+	CurrentReplicas     int32
+	DesiredReplicas     int32
+}
+
+// normalizeDesiredReplicas takes the metrics desired replicas value and normalizes it:
+// 1. It applies the basic conditions (i.e. < maxReplicas, > minReplicas, etc...)
+// 2. It applies the scale up/down limits from the hpaSpec.Constraints (i.e. add no more than 4 pods)
+// 3. It applies the constraints period (i.e. add no more than 4 pods per minute)
+// 4. It applies the constraints delay (i.e. add no more than 4 pods per minute, and pick the smallest recommendation during last 5 minutes)
+func (a *HorizontalController) normalizeDesiredReplicas(hpa *autoscalingv2.HorizontalPodAutoscaler, key string, currentReplicas int32, prenormalizedDesiredReplicas int32) int32 {
+	normalizationArg := NormalizationArg{
+		Key:                 key,
+		ScaleUpConstraint:   generateHPAScaleUpConstraint(hpa.Spec.Constraints),
+		ScaleDownConstraint: generateHPAScaleDownConstraint(hpa.Spec.Constraints),
+		MinReplicas:         hpa.Spec.MinReplicas,
+		MaxReplicas:         hpa.Spec.MaxReplicas,
+		CurrentReplicas:     currentReplicas,
+		DesiredReplicas:     prenormalizedDesiredReplicas}
+	stabilizedRecommendation := a.stabilizeRecommendation(normalizationArg)
+	if stabilizedRecommendation != prenormalizedDesiredReplicas {
+		setCondition(hpa, autoscalingv2.AbleToScale, v1.ConditionTrue, "ScaleDownStabilized", "recent recommendations were higher than current one, applying the highest recent recommendation")
+	} else {
+		setCondition(hpa, autoscalingv2.AbleToScale, v1.ConditionTrue, "ReadyForNewScale", "recommended size matches current size")
+	}
+	desiredReplicas, condition, reason := a.convertDesiredReplicasWithConstraintRate(normalizationArg)
+	if desiredReplicas == stabilizedRecommendation {
+		setCondition(hpa, autoscalingv2.ScalingLimited, v1.ConditionFalse, condition, reason)
+	} else {
+		setCondition(hpa, autoscalingv2.ScalingLimited, v1.ConditionTrue, condition, reason)
+	}
+
+	return desiredReplicas
+}
+
+// getScaleUpChangePerPeriod returns number of pods added during the constrate rate period
+func getScaleUpChangePerPeriod(scaleUpEvents map[string][]timestampedScaleEvent, args NormalizationArg) int32 {
+	cutoff := time.Now().Add(-(time.Duration)(*args.ScaleUpConstraint.Rate.PeriodSeconds))
+	var replicas int32
+	for _, rec := range scaleUpEvents[args.Key] {
+		if rec.timestamp.Before(cutoff) {
+			// outdated scale event should be erased for later usage
+			// We can't do it during storing events as at that point we don't know the PeriodSeconds
+			rec.outdated = true
+		} else {
+			replicas += rec.replicaChange
+		}
+	}
+	return replicas
+}
+
+// getScaleDownChangePerPeriod returns number of pods removed during the constrate rate period
+func getScaleDownChangePerPeriod(scaleDownEvents map[string][]timestampedScaleEvent, args NormalizationArg) int32 {
```

Comment:
`getScaleUpChangePerPeriod()` and `getScaleDownChangePerPeriod()` are identical and I suggest to use a single function instead. For example, `getScaleChangePerPeriod()`.

If you want to keep 2 functions with meaningful names, consider  introducing a 3rd method with a generic logic.

Reply:
I keep them separate as I haven't implemented filling the "scaleUpEvents" and "scaleDownEvents" in the HorizontalController. And that might require additional changes.
So I keep it as it is for now.